### PR TITLE
Get rid of calling get_scale() by patching the step method of optimizer.

### PR DESCRIPTION
**Pull request by yuxinyuan**
Follow up of #1714 . This avoids constantly synchronization between cpu and gpu

### CODE REVIEW

File: src/accelerate/optimizer.py [hunk 1]
```diff
@@ -161,7 +165,26 @@ def step_was_skipped(self):
...
-        return self.__dict__.copy()
+        ret = {}
+        for k, v in self.__dict__.items():
+            if k not in [
+                "_accelerate_num_step_called",
+                "_optimizer_original_step_method",
+                "_optimizer_patched_step_method",
+            ]:
+                ret[k] = v
+        return ret
```
Comment (sgugger) [inline src/accelerate/optimizer.py:176]: A dict comprehension here would be more efficient:
```py
ret = {k: v for k, v in self.__dict__.items() if k not in ...}`
```

**Review by sgugger — APPROVED** [review]
Very clever! I left one comment and I think the `_accelerate_num_step_called` should become a bool called `_accelerate_step_called` since it only has two values with care about (0 or 1), but love that workaround!

File: src/accelerate/optimizer.py [hunk 2]
```diff
@@ -161,7 +165,26 @@ def step_was_skipped(self):
...
 
     def __setstate__(self, state):
         self.__dict__.update(state)
+        self._accelerate_num_step_called = 0
+        self._optimizer_original_step_method = self.optimizer.step
+        self._optimizer_patched_step_method = patch_optimizer_step(self, self.optimizer.step)
+
+
+def patch_optimizer_step(accelerated_optimizer: "AcceleratedOptimizer", method):
```
Comment (muellerzr) [inline src/accelerate/optimizer.py:185]: We can use the class directly for typehints

```suggestion
def patch_optimizer_step(accelerated_optimizer: AcceleratedOptimizer, method):
```

**Review by muellerzr — APPROVED** [review]
This is an excellent workaround! Great work and exceedingly clever :) What Sylvain said, and I just made a small nit for typehinting, as we can use `AcceleratedOptimizer` directly here

File: src/accelerate/optimizer.py (hunk 2 above)
Comment (yuxinyuan) [reply src/accelerate/optimizer.py]: I update this in the latest commit

**Review by muellerzr — APPROVED** [review]

**Comment by sgugger** [comment]
Great! Can you just do the changes on `_accelerate_num_step_called` (see comment above) and we can merge this?

**Comment by yuxinyuan** [comment]
@sgugger oops missed that comment

**Review by sgugger — APPROVED** [review]

**Comment by muellerzr** [comment]
Looks like this is breaking when using deepspeed: https://github.com/huggingface/transformers/actions/runs/5565434638/jobs/10165822075

cc @pacman100 for any guidance?

**Comment by yuxinyuan** [comment]
@muellerzr Seems to me that deepspeed is returning a non-optimizer object and, when we wrap this object with ``AcceleratedOptimizer``, an error will pop up.

I guess a simple way of fixing this is to do the following:
```
if self.scalar is not None:
    self._accelerate_step_called = False
    self._optimizer_original_step_method = self.optimizer.step
    self._optimizer_patched_step_method = patch_optimizer_step(self, self.optimizer.step)
```
Because we only need patching when we are using native amp.

**Pull request merged**